brewtils/decorators.py

```diff
@@ -448,49 +448,57 @@ def _generate_params_from_function(func):
     return parameters_to_return
 
 
-def _generate_nested_params(model_class):
-    """Generates Nested Parameters from a Model Class"""
+def _generate_nested_params(parameter_list):
+    """Generate nested parameters from a list of Parameters
+
+    This function will take a list of Parameters and will return a new list of "real"
+    Parameters.
+
+    The main thing this does is ensure the choices specification is correct for all
+    Parameters in the tree.
+    """
     parameters_to_return = []
-    for parameter_definition in model_class.parameters:
-        key = parameter_definition.key
-        parameter_type = parameter_definition.type
-        multi = parameter_definition.multi
-        display_name = parameter_definition.display_name
-        optional = parameter_definition.optional
-        default = parameter_definition.default
-        description = parameter_definition.description
-        nullable = parameter_definition.nullable
-        maximum = parameter_definition.maximum
-        minimum = parameter_definition.minimum
-        regex = parameter_definition.regex
-        type_info = parameter_definition.type_info
-
-        choices = _format_choices(parameter_definition.choices)
-
-        nested_parameters = []
-        if parameter_definition.parameters:
-            parameter_type = "Dictionary"
-            for nested_class in parameter_definition.parameters:
-                nested_parameters = _generate_nested_params(nested_class)
 
-        parameters_to_return.append(
-            Parameter(
-                key=key,
-                type=parameter_type,
-                multi=multi,
-                display_name=display_name,
-                optional=optional,
-                default=default,
-                description=description,
-                choices=choices,
-                parameters=nested_parameters,
-                nullable=nullable,
-                maximum=maximum,
-                minimum=minimum,
-                regex=regex,
-                type_info=type_info,
+    for param in parameter_list:
+
+        # This is already a Parameter. Only really need to interpret the choices
+        # definition and recurse down into nested Parameters
+        if isinstance(param, Parameter):
+            new_param = Parameter(
+                key=param.key,
+                type=param.type,
+                multi=param.multi,
+                display_name=param.display_name,
+                optional=param.optional,
+                default=param.default,
+                description=param.description,
+                choices=_format_choices(param.choices),
+                nullable=param.nullable,
+                maximum=param.maximum,
+                minimum=param.minimum,
+                regex=param.regex,
+                type_info=param.type_info,
```

Comment:
Since everything maps, minus choices. Could we do something like this instead?

```python
param.choices = _format_choices(param.choices)
new_param = Parameter(**param)
```

Reply:
Not really, `param` here is already an object, not a dict. We *could* do some sort of copy constructor type thing, but I'm not sure it's worth the effort.

I'm also not sure that creating a new Parameter object is the right thing to do in general, but I didn't want to change too much in this PR. There's another issue (beer-garden/beer-garden#777) that'll be sort of a more general cleanup of this code, and I think that'll be something to look at there.